provides.py

```diff
@@ -14,16 +14,29 @@
 from charms.reactive import RelationBase
 from charms.reactive import hook
 from charms.reactive import scopes
+from charmhelpers.core.hookenv import is_leader
+from charmhelpers.core.hookenv import config
+from charmhelpers.core import unitdata
 
 
 class EtcdProvider(RelationBase):
     scope = scopes.GLOBAL
 
     @hook('{provides:etcd}-relation-{joined,changed}')
-    def changed(self):
-        pass
+    def joined_or_changed(self):
+        if is_leader():
+            cstring = self.connection_string()
+            self.set_remote(data={'connection_string': cstring})
 
-    @hook('{provides:etcd}-relation-{broken, departed}')
-    def broken(self):
-        self.remove_state('{relation_name}.available')
-        self.remove_state('{relation_name}.connected')
+    def connection_string(self):
+        if is_leader():
+            db = unitdata.kv()
+            cluster_data = db.get('etcd.cluster_data')
+            connection_string = ""
+            for u in cluster_data:
+                connection_string += ",http://{}:{}".format(cluster_data[u]['private_address'], config('port'))  # noqa
+            return connection_string.lstrip(',')
```

Comment:
This method is making hard assertions about how the charm layer should implement gathering the data for the connection string.  It would be better if your API for this was to have the charm pass in the connection info in response to the relation being joined.  E.g.:

```python
@hook('{provides:etcd}-relation-joined')
def joined(self):
    self.set_state('{relation_name}.joined')
    # document that charm layer should respond to this state and call provide_connection_info

def provide_connection_info(self, hosts, port):
    self.set_remote('connection_string', ','.join(['http://{}:{}'.format(host, port) for host in hosts])
```